### PR TITLE
[Lang] Cuda caching allocator for ndarray 1/n

### PR DESCRIPTION
**Pull request by qiao-bo**
Related issue = #3634 
This is the first step towards ndarray memory recycling. We add a simple caching mechanism to reuse CUDA buffers of same size.

### CODE REVIEW

File: taichi/program/ndarray.cpp [hunk 1]
```diff
@@ -15,18 +15,22 @@ Ndarray::Ndarray(Program *prog,
...
+      nelement_ * element_size_, prog->result_buffer);
 
-  data_ptr_ = prog_impl->get_ndarray_alloc_info_ptr(ndarray_alloc_);
+  data_ptr_ = prog_impl_->get_ndarray_alloc_info_ptr(ndarray_alloc_);
 #else
   TI_ERROR("Llvm disabled");
 #endif
 }
 
+void Ndarray::release() {
```
Comment (ailzhang) [inline taichi/program/ndarray.cpp:30]: after the opengl ndarray pr is merged, this should be moved to ndarray destructor?

File: python/taichi/lang/_ndarray.py [hunk 2]
```diff
@@ -97,6 +97,10 @@ def __getitem__(self, key):
         """
         raise NotImplementedError()
 
+    @python_scope
+    def release(self):
```
Comment (ailzhang) [inline python/taichi/lang/_ndarray.py:101]: IIUC we don't really need a explicit release, ndarray should follow python convention to destruct itself when it's EOL. (which calls into the cpp destructor)

File: taichi/backends/cuda/cuda_device.cpp [hunk 3]
```diff
@@ -38,6 +38,8 @@ DeviceAllocation CudaDevice::allocate_memory_runtime(const AllocParams &params,
   AllocInfo info;
   if (params.host_read || params.host_write) {
     TI_NOT_IMPLEMENTED
+  } else if (params.use_cached && ccalloc.find_block(params.size)) {
```
Comment (ailzhang) [inline taichi/backends/cuda/cuda_device.cpp:41]: IMHO an easier logic here is 
- if use_cached, ccalloc.allocate() // could reuse or allocate new memory
- else: call allocate_llvm_runtime_memory_jit

File: taichi/program/ndarray.cpp (hunk 1 above)
Comment (qiao-bo) [reply taichi/program/ndarray.cpp:30]: yes, will do, thx

File: taichi/backends/cpu/cpu_device.cpp [hunk 4]
```diff
@@ -26,14 +27,12 @@ DeviceAllocation CpuDevice::allocate_memory(const AllocParams &params) {
...
-                                                    LLVMRuntime *runtime,
-                                                    uint64 *result_buffer) {
+DeviceAllocation CpuDevice::allocate_memory_runtime(
+    AllocParamsLlvmRuntime &params) {
   AllocInfo info;
-  info.ptr = allocate_llvm_runtime_memory_jit(runtime_jit, runtime, params.size,
-                                              result_buffer);
+  info.ptr = allocate_llvm_runtime_memory_jit(params);
   info.size = params.size;
+  info.use_cached = params.use_cached;  // TODO add cpu caching allocator
```
Comment (k-ye) [inline taichi/backends/cpu/cpu_device.cpp:35]: Nit for here and below: The widely used format is `// TODO: ...`, and is written on a separate line

File: taichi/backends/device.h [hunk 5]
```diff
@@ -389,19 +389,23 @@ class Device {
...
     TI_NOT_IMPLEMENTED
   }
+
   virtual void dealloc_memory(DeviceAllocation handle) = 0;
 
-  uint64_t *allocate_llvm_runtime_memory_jit(JITModule *runtime_jit,
-                                             LLVMRuntime *runtime,
-                                             size_t size,
-                                             uint64 *result_buffer);
+  uint64_t *allocate_llvm_runtime_memory_jit(AllocParamsLlvmRuntime &params);
```
Comment (k-ye) [inline taichi/backends/device.h:408]: Can we make `params` const?

File: taichi/backends/device.h [hunk 6]
```diff
@@ -389,19 +389,23 @@ class Device {
     AllocUsage usage{AllocUsage::Storage};
   };
 
+  struct AllocParamsLlvmRuntime : AllocParams {
```
Comment (k-ye) [inline taichi/backends/device.h:392]:
```suggestion
  struct LlvmRuntimeAllocParams : AllocParams {
```

File: taichi/backends/cuda/cuda_caching_allocator.h [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace taichi {
+namespace lang {
+namespace cuda {
+
+class CudaCachingAllocator {
+ public:
+  CudaCachingAllocator(Device *device);
+
+  void release(size_t sz, uint64_t *ptr);
```
Comment (k-ye) [inline taichi/backends/cuda/cuda_caching_allocator.h:16]: nit: allocate() comes before deallocate()

File: taichi/program/compile_config.h [hunk 8]
```diff
@@ -42,6 +42,7 @@ struct CompileConfig {
   bool make_block_local;
   bool detect_read_only;
   bool ndarray_use_torch;
+  bool ndarray_use_cached_allocator;
```
Comment (k-ye) [inline taichi/program/compile_config.h:45]: Do we need this additional flag?

**Review by k-ye — COMMENTED** [review]
Thanks! I have a few nits

File: taichi/program/compile_config.h (hunk 8 above)
Comment (qiao-bo) [reply taichi/program/compile_config.h:45]: Just to have an option to turn the caching behavior off. It might be useful later for profiling.

File: taichi/backends/device.h (hunk 5 above)
Comment (qiao-bo) [reply taichi/backends/device.h:408]: ehm, this struct has the runtime_jit and result buffer that prevent const. Maybe a better name is `LlvmRuntimeAllocContext`? which can contain a sub struct that is of params const?

File: taichi/backends/cuda/cuda_caching_allocator.h [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+namespace lang {
+namespace cuda {
+
+class CudaCachingAllocator {
+ public:
+  CudaCachingAllocator(Device *device);
+
+  uint64_t *allocate(const Device::LlvmRuntimeAllocParams &params);
+  void release(size_t sz, uint64_t *ptr);
+  bool find_block(size_t sz) const;
```
Comment (k-ye) [inline taichi/backends/cuda/cuda_caching_allocator.h:18]: super nit: this can be private?

**Review by k-ye — APPROVED** [review]
LGTM!

File: taichi/program/ndarray.cpp [hunk 10]
```diff
@@ -8,22 +8,22 @@ namespace lang {
...
       nelement_(std::accumulate(std::begin(shape),
                                 std::end(shape),
                                 1,
                                 std::multiplies<>())),
       element_size_(data_type_size(dtype)),
-      device_(prog->get_device_shared()) {
-  ndarray_alloc_ = prog->allocate_memory_ndarray(nelement_ * element_size_,
-                                                 prog->result_buffer);
-
+      device_(prog_->get_device_shared()) {
```
Comment (ailzhang) [inline taichi/program/ndarray.cpp:20]: ditto, changes here don't seem necessary?

File: taichi/llvm/llvm_program.cpp [hunk 11]
```diff
@@ -574,11 +574,18 @@ DeviceAllocation LlvmProgramImpl::allocate_memory_ndarray(
     tlctx = llvm_context_host.get();
   }
 
-  Device::AllocParams device_buffer_alloc_params;
+  Device::LlvmRuntimeAllocParams device_buffer_alloc_params;
```
Comment (ailzhang) [inline taichi/llvm/llvm_program.cpp:577]: Can we initialize those values at construction time using `{}`?

File: taichi/backends/cuda/cuda_device.h [hunk 12]
```diff
@@ -80,19 +81,19 @@ class CudaDevice : public Device {
...
 
   ~CudaDevice() override{};
 
   DeviceAllocation allocate_memory(const AllocParams &params) override;
-  DeviceAllocation allocate_memory_runtime(const AllocParams &params,
-                                           JITModule *runtime_jit_module,
-                                           LLVMRuntime *runtime,
-                                           uint64 *result_buffer) override;
+  DeviceAllocation allocate_memory_runtime(
+      const LlvmRuntimeAllocParams &params) override;
```
Comment (ailzhang) [inline taichi/backends/cuda/cuda_device.h:93]: Ideally packing things into struct `LlvmRuntimeAllocParams` should be a separate PR which can be approved pretty quicky ;)

File: taichi/backends/cuda/cuda_device.h [hunk 13]
```diff
@@ -119,6 +120,7 @@ class CudaDevice : public Device {
       TI_ERROR("invalid DeviceAllocation");
     }
   }
+  std::unique_ptr<CudaCachingAllocator> caching_allocator_{nullptr};
```
Comment (ailzhang) [inline taichi/backends/cuda/cuda_device.h:123]: private?

File: taichi/program/compile_config.h (hunk 8 above)
Comment (ailzhang) [reply taichi/program/compile_config.h:45]: IMHO caching is strictly better than the one without, so we probably can just show some perf comparison in this PR and then save us from adding a new compile config (and the branches). But I'm pretty open on this. :D

File: taichi/backends/cuda/cuda_caching_allocator.cpp [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    ret = blk->second;
+    mem_blocks_.erase(blk);
+  } else {
+    ret = device_->allocate_llvm_runtime_memory_jit(params);
+  }
+  return ret;
+}
+
+void CudaCachingAllocator::release(size_t sz, uint64_t *ptr) {
+  mem_blocks_.insert(std::pair<size_t, uint64_t *>(sz, ptr));
```
Comment (ailzhang) [inline taichi/backends/cuda/cuda_caching_allocator.cpp:24]: does `{sz, ptr}` work? ;)

**Review by ailzhang — COMMENTED** [review]
Nice work!

File: taichi/backends/cuda/cuda_device.h (hunk 13 above)
Comment (qiao-bo) [reply taichi/backends/cuda/cuda_device.h:123]: it is already private ;)

File: taichi/program/ndarray.h [hunk 15]
```diff
@@ -34,6 +34,7 @@ class Ndarray {
   ~Ndarray();
 
  private:
+  Program *prog_{nullptr};
```
[the remark this answers is not in the document]
Comment (qiao-bo) [reply taichi/program/ndarray.h:37]: right, forgot to delete from previous change. thanks!
